### ISSUE TITLE
Surface author_api_key_id (or actual API key)

### ISSUE THREAD

**Issue by KetchupBomb** [shlinkio/shlink#1059]
#### Summary

tl;dr - The table:field `short_urls:author_api_key_id` seems to record the `api_keys:id` field, but this information isn't surfaced in the CLI or web clients. It would be nice to be able to audit this information without requiring database access.

Specifically, I would love to be able to see which of my API users are generating links, reaping old and unused keys, while also being able to compare usage across my users.

---

```shell
mysql> DESC short_urls;
+----------------------------+---------------------+------+-----+---------+----------------+
| Field                      | Type                | Null | Key | Default | Extra          |
+----------------------------+---------------------+------+-----+---------+----------------+
| id                         | bigint(20) unsigned | NO   | PRI | NULL    | auto_increment |
| domain_id                  | bigint(20) unsigned | YES  | MUL | NULL    |                |
| author_api_key_id          | bigint(20) unsigned | YES  | MUL | NULL    |                |    // This is likely api_keys:id
| original_url               | varchar(2048)       | NO   |     | NULL    |                |
| short_code                 | varchar(255)        | NO   | MUL | NULL    |                |
| date_created               | datetime            | NO   |     | NULL    |                |
| valid_since                | datetime            | YES  |     | NULL    |                |
| valid_until                | datetime            | YES  |     | NULL    |                |
| max_visits                 | int(11)             | YES  |     | NULL    |                |
| import_source              | varchar(255)        | YES  |     | NULL    |                |
| import_original_short_code | varchar(255)        | YES  |     | NULL    |                |
| title                      | varchar(512)        | YES  |     | NULL    |                |
| title_was_auto_resolved    | tinyint(1)          | NO   |     | 0       |                |
+----------------------------+---------------------+------+-----+---------+----------------+
13 rows in set (0.01 sec)



Database changed
mysql> DESC api_keys;
+-----------------+---------------------+------+-----+---------+----------------+
| Field           | Type                | Null | Key | Default | Extra          |
+-----------------+---------------------+------+-----+---------+----------------+
| id              | bigint(20) unsigned | NO   | PRI | NULL    | auto_increment |
| key             | varchar(255)        | NO   | UNI | NULL    |                |
| expiration_date | datetime            | YES  |     | NULL    |                |
| enabled         | tinyint(1)          | NO   |     | NULL    |                |
+-----------------+---------------------+------+-----+---------+----------------+
4 rows in set (0.00 sec)
```

What are your thoughts on surfacing this info in the CLI with another flag?

```shell
$ shlink short-url:list --help | grep -iE 'api|key' | wc -l
0
```

Rather, is there a reason this isn't already exposed?

**Comment by acelaya** [comment]
There's no particular reason for not exposing this in the CLI command.

The field was initially added to support restricting access only to the resources created with one particular API key, but the intention was to expose that info at some point. See the last paragraph here https://github.com/shlinkio/shlink/issues/795#issuecomment-759431445

So yeah, it's something useful and it can be added. Si you want to contribute the change?

**Comment by KetchupBomb** [comment]
I see. Thanks for the history.

Yeah, I'll give it a go. 👌

**Comment by KetchupBomb** [comment]
@acelaya, I've been reviewing the existing code to see where I can/should hook in.

I've found [`class ShortUrl`](https://github.com/shlinkio/shlink/blob/develop/module/Core/src/Entity/ShortUrl.php), which appears to be a logical object, representing various aspects of a short URL, including resolution of database relations. For instance, the database table `short_urls` doesn't include tags, but rather the table `short_urls_in_tags` contains the mapping of short URL IDs to tag IDs, and the mentioned class attempts to perform this resolution, associating PHP objects of short URLs with Tags.

`ShortUrl` already has an [`authorApiKey`](https://github.com/shlinkio/shlink/blob/develop/module/Core/src/Entity/ShortUrl.php#L76) attribute, though it isn't resolved -- it's only defined when meta data is passed in (which it seems only occurs on short URL creation). I was thinking to declare an additional [`ShortUrlRelationResolverInterface`](https://github.com/shlinkio/shlink/blob/develop/module/Core/src/ShortUrl/Resolver/ShortUrlRelationResolverInterface.php) function for resolution, but when reviewing [`PersistenceShortUrlRelationResolver`](https://github.com/shlinkio/shlink/blob/e30c9c86ff1e05217284cf2e0070b493583ac67d/module/Core/src/ShortUrl/Resolver/PersistenceShortUrlRelationResolver.php) and [`SimpleShortUrlRelationResolver`](https://github.com/shlinkio/shlink/blob/e30c9c86ff1e05217284cf2e0070b493583ac67d/module/Core/src/ShortUrl/Resolver/SimpleShortUrlRelationResolver.php) classes, along with the functions to resolve [`Domain`](https://github.com/shlinkio/shlink/blob/d426dbc684ec7f9a81eaaf43ce189e23355ed61d/module/Core/src/Entity/Domain.php) and [`Tag`](https://github.com/shlinkio/shlink/blob/d426dbc684ec7f9a81eaaf43ce189e23355ed61d/module/Core/src/Entity/Tag.php) objects, I don't understand how/where to specify the necessary database SELECT & (LEFT) JOIN statements to get the relevant data for an `ApiKey` Entity object.

It looks like there may be a need for an `ApiKeyRepository` class, similar to how there are [`ShortUrlRepository`](https://github.com/shlinkio/shlink/blob/c931874bacdfb89a82bb3d6a5ba4bacc1dc2a260/module/Core/src/Repository/ShortUrlRepository.php) and [`TagRepository`](https://github.com/shlinkio/shlink/blob/c931874bacdfb89a82bb3d6a5ba4bacc1dc2a260/module/Core/src/Repository/TagRepository.php) classes. But I'm having a hard time understanding the Doctrine code, specifically how to indicate which table(s) to use -- I can see how [a deletion example potentially uses a class name](https://github.com/shlinkio/shlink/blob/c931874bacdfb89a82bb3d6a5ba4bacc1dc2a260/module/Core/src/Repository/TagRepository.php#L26), but then I don't see how tables are specified [when performing a join](https://github.com/shlinkio/shlink/blob/c931874bacdfb89a82bb3d6a5ba4bacc1dc2a260/module/Core/src/Repository/TagRepository.php#L38).

---

Two main questions:
1. Based on my above description, do you think my logic of where to add, and my opinion of resolving the `ShortUrl->authorApiKey`, are the right moves?
2. Assume the above point is on track, can you help me understand how table resolution works in this case?

**Comment by acelaya** [comment]
Hey @KetchupBomb 

You actually don't need all that, it's far simpler.

You are right that the `ShortUrl` is the domain entity that represents a short URL. The logic to map its properties to the persistence model is done by doctrine, using this configuration https://github.com/shlinkio/shlink/blob/develop/module/Core/config/entities-mappings/Shlinkio.Shlink.Core.Entity.ShortUrl.php

So, you just need to follow these steps:

* Expose the `authorApiKey` through a getter.
* Improve the `ListShortUrlsCommand` so that it reads the API key using that getter, and prints its key and/or name if it's not null.
* Fix/add tests where applicable.

For the second step, I'd suggest having a private method that has the logic to represent an API key as string:

```php
private function apiKeyToString(?ApiKey apiKey): string
{
    if ($apiKey === null) {
        return '-';
    }

    $name = $apiKey->name();
    $key = $apiKey->toString();
    if (empty($name)) {
        return $key;
    }

    return sprintf('%s (%s)', $key, $title);
}
```

I'm not sure if the title should be printed as a replacement of the key when it is not empty, or in combination with the key as in the example. I'm open to suggestions there.

**Comment by acelaya** [comment]
Closed by #1064
